### PR TITLE
KAFKA-4293 - improve ByteBufferMessageSet.deepIterator() performance by relying on underlying stream's available() implementation

### PR DESCRIPTION
**Pull request by radai-rosenblatt**
also:
provided better available() for ByteBufferInputStream
provided better available() for KafkaLZ4BlockInputStream
added KafkaGZIPInputStream with a better available()
fixed KafkaLZ4BlockOutputStream.close() to properly flush

Signed-off-by: radai-rosenblatt radai.rosenblatt@gmail.com

### CODE REVIEW

**Comment by radai-rosenblatt** [comment]
The long-term correct solution to the bug is still a wire format change.
However, shorter term this PR makes the code rely on InputStream.available() to detect EOF without catching an exception. this works because for this particular case the reads are all memory-backed (as opposed to directly from socket) and so all input is at hand - all that's required is careful implementation of available().

**Comment by ijuma** [comment]
Have we verified that this change makes an actual difference outside the profiler? I ask because many profilers are known to suffer from safepoint bias. Reference: http://psy-lob-saw.blogspot.co.uk/2016/02/why-most-sampling-java-profilers-are.html

File: clients/src/main/java/org/apache/kafka/common/record/KafkaGZIPInputStream.java [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+public class KafkaGZIPInputStream extends GZIPInputStream {
+
+    public KafkaGZIPInputStream(InputStream in) throws IOException {
+        super(in);
+    }
+
+    @Override
+    public int available() throws IOException {
+        int superResult = super.available();
+        if (superResult < 1) {
```
Comment (jjkoshy) [inline clients/src/main/java/org/apache/kafka/common/record/KafkaGZIPInputStream.java:38]: While we should have the call to `super` above, do we really need this given that the default `available` implementation in GZIP is not very helpful? i.e.,

```
super.available();
return inf.finished() ? 0 : 1;
```

**Comment by radai-rosenblatt** [comment]
@ijuma - thank you for the article, it was fascinating, and it may indeed be that our profiling runs were affected.

to gauge the effect of this change i used the following "micro-benchmark" (quoted because i didnt bother with a full JMH one):

``` java
@Test
def perfTest(): Unit = {
   val batchSize = 1
   var nanosTotal: Long = 0
   var runsTotal: Int = 0
   (1 to 1000000).foreach(i => {
     val messages = (1 to batchSize).map(m => {
       new Message((s"message num $m" + System.currentTimeMillis()).getBytes)
     })
     val messageSet = new ByteBufferMessageSet(GZIPCompressionCodec, messages: _*)
     var count = 0
     val start = System.nanoTime()
     messageSet.foreach((offset: MessageAndOffset) => {
       count += 1
     })
     val elapsed = System.nanoTime() - start
     assertEquals(batchSize, count)
     if (i > 100000) { //warm up iterations
       nanosTotal += elapsed
       runsTotal += 1
     }
   })
   val avg = nanosTotal / (runsTotal * 1000000.0)
   println(s"$avg ms on avg")
}
```

comparing the avg per-betch iteration time between vanilla Kafka and my patch shows it highly depends on the batch parameters.

The most extreme case is small batches and fast compression - batch size 1 + snappy show a ~x4 time improvement. this would be typical of environments where kafka would be used for real time messaging (tiny batches, fast compression).

On the other extreme (batch size 10000 + gzip) there's no noticeable difference.

Also no difference in case no compression was used.

**Comment by radai-rosenblatt** [comment]
@ijuma - have you guys had time to evaluate this?

**Comment by ijuma** [comment]
@radai-rosenblatt I haven't reviewed the patch in detail, but it seems like it might be worth including. @junrao, what do you think?

File: core/src/test/scala/unit/kafka/message/ByteBufferMessageSetTest.scala [hunk 2]
```diff
@@ -422,6 +422,22 @@ class ByteBufferMessageSetTest extends BaseMessageSetTestCases {
     checkWriteFullyToWithMessageSet(createMessageSet(messages))
   }
 
+  @Test
+  def testIterationOverCompressedSet(): Unit = {
+    val messageCount = 100
+    val messages = (1 to messageCount).map(m => {
+      new Message(s"message num $m".getBytes)
+    })
+    var messageSet = new ByteBufferMessageSet(GZIPCompressionCodec, messages:_*)
```
Comment (jjkoshy) [inline core/src/test/scala/unit/kafka/message/ByteBufferMessageSetTest.scala:431]: Can you replace these with a parameterized test?

File: clients/src/test/java/org/apache/kafka/common/record/CompressionCodecsTest.java [hunk 3]
```diff
@@ -0,0 +1,217 @@
...
+            },
+            new OutputStreamWrapper() {
+                @Override
+                public OutputStream wrap(OutputStream underlying) throws Exception {
+                    return new GZIPOutputStream(underlying); //use vanilla
+                }
+            });
+    }
+
+    @Test
```
Comment (jjkoshy) [inline clients/src/test/java/org/apache/kafka/common/record/CompressionCodecsTest.java:109]: All these tests could also be parameterized - would be a lot less code, but would likely need reflection to look up the constructor.

**Review by jjkoshy — CHANGES_REQUESTED** [review]

**Comment by jjkoshy** [comment]
@ijuma @junrao I think the patch makes sense - at the very least it avoids the use of exceptions for flow-control. That said, our original toy benchmark was flawed so the savings are not as dramatic as we hoped. Still, it does show some improvement as @radai-rosenblatt noted above.

File: clients/src/main/java/org/apache/kafka/common/record/KafkaGZIPInputStream.java (hunk 1 above)
Comment (radai-rosenblatt) [reply clients/src/main/java/org/apache/kafka/common/record/KafkaGZIPInputStream.java:38]: its the "generally correct" thing to do. given the parent method is extremely short im betting the JIT just inlines it away.

File: core/src/test/scala/unit/kafka/message/ByteBufferMessageSetTest.scala (hunk 2 above)
Comment (radai-rosenblatt) [reply core/src/test/scala/unit/kafka/message/ByteBufferMessageSetTest.scala:431]: will do

File: clients/src/test/java/org/apache/kafka/common/record/CompressionCodecsTest.java (hunk 3 above)
Comment (radai-rosenblatt) [reply clients/src/test/java/org/apache/kafka/common/record/CompressionCodecsTest.java:109]: it wont be much less code - the actual test logic is already factored out to a common method, and the constructor signature for the various streams is not completely identical, so no easy reflexive solution. java 8 closures are the only big win i can see here, and kafka is (unfortunately) still java 7

**Comment by ijuma** [comment]
Had a closer look at the javadoc for `available`:

``` java
* Returns an estimate of the number of bytes that can be read (or
     * skipped over) from this input stream without blocking by the next
     * invocation of a method for this input stream. The next invocation
     * might be the same thread or another thread.  A single read or skip of this
     * many bytes will not block, but may read or skip fewer bytes.
     *
     * <p> Note that while some implementations of {@code InputStream} will return
     * the total number of bytes in the stream, many will not.  It is
     * never correct to use the return value of this method to allocate
     * a buffer intended to hold all data in this stream.
     *
     * <p> A subclass' implementation of this method may choose to throw an
     * {@link IOException} if this input stream has been closed by
     * invoking the {@link #close()} method.
     *
     * <p> The {@code available} method for class {@code InputStream} always
     * returns {@code 0}.
     *
     * <p> This method should be overridden by subclasses.
     *
     * @return     an estimate of the number of bytes that can be read (or skipped
     *             over) from this input stream without blocking or {@code 0} when
     *             it reaches the end of the input stream.
     * @exception  IOException if an I/O error occurs.
```

So even though I think it makes sense to avoid the exceptions, it seems a bit brittle to rely on that method as an alternative to reading until the end. Yes, we are overriding all the implementations to tighten the contract, but it seems possible that someone would add a new implementation that satisfies the contract and breaks the current expectation.

It would seem that it would be better to have a separate method (and an interface perhaps) that does what we need?

**Comment by jjkoshy** [comment]
Yes we did notice that detail and it's a valid concern. Your suggestion on having a more explicit interface makes sense. The codecs extend a sealed trait and we could add some protection (in the tests) that all codecs are tested.

**Comment by radai-rosenblatt** [comment]
....and this is exactly why I think the long term correct solution is a wire format change - declare sizes and counts up front.

@ijuma - as InputStream is an abstract class rather than an interface I dont see a way of doing what you proposed that does not involve multiple inheritance?

**Comment by ijuma** [comment]
~~An abstract class that extends InputStream may be a better option in that case~~. Actually that probably doesn't work. Let me check the diff more carefully.

**Comment by radai-rosenblatt** [comment]
`KafkaGZIPInputStream extends GZIPInputStream`  <--- i have to do this to get access to a protected field :-(

File: core/src/test/scala/unit/kafka/message/ByteBufferMessageSetTest.scala (hunk 2 above)
Comment (radai-rosenblatt) [reply core/src/test/scala/unit/kafka/message/ByteBufferMessageSetTest.scala:431]: after actually doing this i realized that there are 13 other test methods in this class, that would runx4 times for no reason if i parameterize the class just for my test. you still want to do this?

**Comment by jjkoshy** [comment]
The other option is to drop this altogether :)
Perhaps the savings aren't sufficient to justify these concerns.

**Comment by radai-rosenblatt** [comment]
the issue with GZIPInputStream has been fixed in jdk9 (which means my class from this patch could go away at some distant point in the future) - http://bugs.java.com/bugdatabase/view_bug.do?bug_id=7031075

**Comment by xvrl** [comment]
@radai-rosenblatt I've submitted the following patch as part of KAFKA-5150 that should accomplish the same thing with a smaller set of changes, in case you are interested
https://github.com/apache/kafka/pull/2967/commits/97e3ce81cea2736232a9527dc6a8cbd29e366ccd

see #2967

**Comment by radai-rosenblatt** [comment]
i'll defer to  @xvrl 'spatch

**Pull request closed without merging**